lib/elixir/lib/module/checker.ex

```diff
@@ -4,98 +4,113 @@ defmodule Module.Checker do
       file: module_map.file,
       module: module_map.module,
       compile_opts: module_map.compile_opts,
-      function: nil
+      function: nil,
+      warnings: []
     }
 
     module_map.definitions
     |> check_definitions(state)
-    |> List.flatten()
+    |> Map.fetch!(:warnings)
     |> merge_warnings()
     |> sort_warnings()
     |> emit_warnings()
   end
 
   defp check_definitions(definitions, state) do
-    Enum.map(definitions, &check_definition(&1, state))
+    Enum.reduce(definitions, state, &check_definition/2)
   end
 
   defp check_definition({function, _kind, meta, clauses}, state) do
-    state = put_file_meta(%{state | function: function}, meta)
-    Enum.map(clauses, &check_clause(&1, state))
+    with_file_meta(%{state | function: function}, meta, fn state ->
```

Comment:
This doesn't need to receive an anonymous function:

```
state = with_file_meta(%{state | function: function}, meta)
Enum.reduce(clauses, state, &check_clause/2)
```

Reply:
We will leak the file meta to other definitions if we don't do this. We can reset the state in `check_definition/2` but it seems better to localize to the function that sets `state.file`.

Reply:
Ah, good catch.